compiler/resolution/src/org/jetbrains/kotlin/resolve/calls/tower/TowerResolver.kt

```diff
@@ -132,28 +142,50 @@ class TowerResolver {
                 }
             }
 
-            lexicalScope.parentsWithSelf.forEach { scope ->
-                if (scope is LexicalScope) {
-                    if (!scope.kind.withLocalDescriptors) {
-                        addLevel(
-                            ScopeBasedTowerLevel(this@createNonLocalLevels, scope),
-                            scope.mayFitForName(name)
-                        )
-                    }
-
-                    getImplicitReceiver(scope)?.let {
-                        addLevel(
-                            MemberScopeTowerLevel(this@createNonLocalLevels, it),
-                            it.mayFitForName(name)
-                        )
-                    }
-                } else {
+            fun addLevelForLexicalScope(scope: LexicalScope) {
+                if (!scope.kind.withLocalDescriptors) {
                     addLevel(
-                        ImportingScopeBasedTowerLevel(this@createNonLocalLevels, scope as ImportingScope),
+                        ScopeBasedTowerLevel(this@createNonLocalLevels, scope),
                         scope.mayFitForName(name)
                     )
                 }
+
+                getImplicitReceiver(scope)?.let {
+                    addLevel(
+                        MemberScopeTowerLevel(this@createNonLocalLevels, it),
+                        it.mayFitForName(name)
+                    )
+                }
+            }
+
+            fun addLevelForContextReceiverGroup(contextReceiversGroup: List<ReceiverValueWithSmartCastInfo>) =
+                addLevel(
+                    ContextReceiversGroupScopeTowerLevel(this@createNonLocalLevels, contextReceiversGroup),
+                    contextReceiversGroup.any { it.mayFitForName(name) }
+                )
+
+            fun addLevelForImportingScope(scope: HierarchicalScope) =
+                addLevel(
+                    ImportingScopeBasedTowerLevel(this@createNonLocalLevels, scope as ImportingScope),
+                    scope.mayFitForName(name)
+                )
+
+            val parentScopes = lexicalScope.parentsWithSelf
```

Comment:
I'd converted it to List, to avoid double traversal

Reply:
Fixed. How does it help to avoid double traversal?